llvm/lib/Target/PowerPC/PPCTargetMachine.cpp

```diff
@@ -500,7 +500,10 @@ void PPCPassConfig::addIRPasses() {
 }
 
 bool PPCPassConfig::addPreISel() {
-  if (EnableGlobalMerge)
+  if ((EnableGlobalMerge.getNumOccurrences() > 0)
+          ? EnableGlobalMerge
+          : (TM->getTargetTriple().isOSAIX() &&
+             getOptLevel() != CodeGenOptLevel::None))
```

Comment:
I think the condition can be simplified:

```suggestion
  if (EnableGlobalMerge
      || (TM->getTargetTriple().isOSAIX() &&
             getOptLevel() != CodeGenOptLevel::None))
```
(formatting is most likely wrong)

Reply:
Hmm, I thought this could be simplified, too.

However, unless I am misunderstanding or I did something wrong, it seems the original code is still needed or else `xcoff-symbolize-data.ll` fails:
```
llvm-project/llvm/test/DebugInfo/Symbolize/XCOFF/xcoff-symbolize-data.ll:38:10: error: CHECK: expected string not found in input
; CHECK: f()::function_global
         ^
<stdin>:16:1: note: scanning from here

^
<stdin>:17:1: note: possible intended match here
L.._MergedGlobals
^
```
I can see in that test that it seems we're still merging (with this code suggestion) since the `_MergedGlobals` is present:
```
$ llvm-symbolizer --obj=/data/amyk/llvm_trunk/build/test/DebugInfo/Symbolize/XCOFF/Output/xcoff-symbolize-data.ll.tmp 'DATA 0x60' 'DATA 0x61' 'DATA 0x64' 'DATA 0X68'    'DATA 0x90' 'DATA 0x94'
bss_global
96 4
/t.cpp:1

bss_global
96 4
/t.cpp:1

data_global
100 4
/t.cpp:2

str
104 4
/t.cpp:4

L.._MergedGlobals
144 8
/t.cpp:13

L.._MergedGlobals
144 8
/t.cpp:12
```

Any opinions regarding this and/or keeping the original code, @chenzheng1030 @redstar?

Reply:
First thought: there is a slight change in the semantics.
The condition `EnableGlobalMerge.getNumOccurrences() > 0` checks how often the option was specified on the command line. If I specify `-ppc-global-merge=false` then the original `if` returns `false` (because the option was specified 1 time, and it's value is `false` but the value of the changed condition depends on being on AIX or not.

If the former is really the intended functionality then IMHO it warrants a big comment (e.g. "Specifying the command line option overrides AIX default"), just to clarify it.

Reply:
Thanks a lot, Kai! I looked into this a bit and I agree. That would be the behaviour that we are after here, so I will keep the condition as is, and add a comment.